Readme.txt

  1. About Dos32cm

   Dos32cm is a kind of Dos64stub variant - it allows to run programs in long
  mode, but, unlike Dos64stub, launches the program in compatibility sub-mode.
   Dos32cm is a so-called stub, a ( small ) program supposed to be added to a
  ( 32-bit PE ) binary by the link step. The stub will be executed when the
  binary is launched in DOS.
  
  During startup, Dos32cm will do:

   - check if the cpu is 64-bit
   - check if the PE image is "acceptable"
   - check if enough XMS memory is available
   - setup IDT and page tables for 64-bit PAE paging
   - read & move the image into extended memory
   - install a small "OS" (int 21h/31h) so one may call
     real-mode DOS functions that don't need pointer translation.
   - reprogram master PIC, so IRQs 00h-07h are mapped to Int 78h-7fh
   - switch to long-mode
   - call the entry point of the loaded 32-bit image

   Dos32cm doesn't require the image to contain base relocations,
  because it should be able to load the image at its prefered load address.
  Dos32cm maps just the memory that the image needs to run, that is:
  
   - image, stack & heap
   - conventional memory region, address 0-0xfffffh
   - IDT, mapped at 0x100000.


  2. Requirements
  
  to run an image with Dos32cm attached one needs:

   - a 64-bit CPU
   - an installed DOS
   - an installed XMS host
   - enough extended memory to load the image

  Tested on real machines and in Qemu.


  3. How to use the Stub?

  The stub is added to a 32-bit binary thru the link step. See the sample
  Makefiles for how to do this with jwlink, MS link or pestub ( a tool borrowed
  from HX ). The image must meet the following requirements:

   - Subsystem is "native"; the image really shouldn't run as Windows app
   - no dll references ("imports") are possible

  The stubs install a tiny subset of the DPMI API. The functions that are
  supported are:
   - int 21h, ah=4Ch: terminate program
   - int 31h, ax=203h: set exception vector BL to CX:EDX ( must be 64-bit! )
   - int 31h, ax=300h: simulate real-mode interrupt BL, EDI=real-mode call
     structure.
  
   One of the samples, Mon32, allows to display memory. It shows how the
  Int21 emulation installed by Dos32cm is supposed to be used and also may
  demonstrate page fault handling if an invalid address is entered.

   Another sample, Mon32x, demonstrates how to switch from compatibility mode
  to 64-bit mode. JWasm v2.17+ has to be used to create the binary, using
  JWasm's -pe option - no idea if there exist linkers that support mixing
  32- and 64-bit code in a PE binary.


  4. Memory Layout
 
  Dos32cm doesn't map the paging tables, and the IDT is always mapped at
  0x100000h. The image itself is mapped at its prefered load address.


  5. License
  
  the source is distributed under the MIT license. See file
  COPYING for details. It was written by Andreas Grech.



History.txt

   15.12.2022:
   - initial


Mon32/MON32.asm

;--- this is a simple monitor program loaded by dos32cm.bin

    .386
    .model flat
    option casemap:none
    option proc:private

lf  equ 10 

;--- define a string
CStr macro text:vararg
local sym
    .const
sym db text,0
    .code
    exitm <offset sym>
endm

    .data?

dwEsp   dd ?    ;stack pointer
address dd ?

;--- keyboard buffer
buffer db 20 dup (?)
lbuffer equ $ - offset buffer

    .code

if 0
	include vioout.inc
	include dprintf.inc
printf textequ <dprintf>
else
    include printf.inc
endif

;--- this is the entry point.
;--- it's called by the stub with registers:
;--- ebx: image base
;--- esp: bottom of (reserved) stack

main proc c

    invoke printf, CStr("Mon32 loaded at %X, esp=%X",lf), ebx, esp
    invoke printf, CStr("cs=%X ss=%X ds=%X es=%X fs=%X gs=%x",lf), cs, ss, ds, es, fs, gs
    invoke printf, CStr("ax=%X bx=%X cx=%X dx=%X si=%X di=%X bp=%X",lf), eax, ebx, ecx, edx, esi, edi, ebp
    call set_exception_handlers
nextcmd:
    invoke printf, CStr("(cmds: a,d,q): ")

    mov ah,1        ;read a key from keyboard with echo
    int 21h
    push offset nextcmd
    mov dwEsp, esp
    push eax
    invoke printf, CStr(lf)
    pop eax
    cmp al,'a'
    jz a_cmd
    cmp al,'d'
    jz d_cmd
    cmp al,'q'
    jz q_cmd
    cmp al,0dh      ;ENTER?
    jz newline
    mov ecx, eax
    invoke printf, CStr("unknown cmd: %c",lf), ecx
newline:
    ret
main endp

;--- get a line of characters
;--- ebx->buffer
;--- esi=size of buffer

getline proc
    xor edi, edi
    dec esi
nextkey:
    mov ah,1
    int 21h
    cmp al,0dh
    jz enter_pressed
    cmp al,08h
    jz backspace_pressed
    mov [ebx+edi],al
    inc edi
    cmp edi,esi
    jnz nextkey
    mov dl,0dh
    mov ah,2
    int 21h
    jmp enter_pressed
backspace_pressed:
    mov dl,20h
    mov ah,2
    int 21h
    cmp edi,0
    jz nextkey
    dec edi
    mov dl,08h
    mov ah,2
    int 21h
    jmp nextkey
enter_pressed:    
    mov byte ptr [ebx+edi],0
    ret
getline endp

;--- enter address (for d cmd)

a_cmd proc
    invoke printf, CStr("enter start address for d cmd: ")
    mov esi, lbuffer
    lea ebx, buffer
    call getline
    and edi,edi        ;at least 1 digit entered?
    jz done
    xor edi,edi
    xor esi,esi
    .while byte ptr [ebx+edi]
        mov al,byte ptr [ebx+edi]
        sub al,'0'
        jc error
        cmp al,9
        jbe @F
        sub al, 27h
        cmp al, 0fh
        ja error
@@:
        movzx eax,al
        shl esi,4
        add esi,eax
        inc edi
    .endw
    mov [address],esi
done:
    invoke printf, CStr(lf)
    ret
error:
    invoke printf, CStr(lf,"%s?",lf), addr buffer
    ret

a_cmd endp

;--- display memory dump

d_cmd proc
    mov esi,[address]
    mov ecx,8
nextline:
    push ecx
    invoke printf, CStr("%08X: "), esi
    mov ecx,16
@@:
    push ecx
    lodsb
    movzx eax, al
    invoke printf, CStr("%02X "), eax
    pop ecx
    loop @B
    invoke printf, CStr(" ")
    sub esi,16
    mov ecx,16
nextc:
    lodsb
    cmp al,20h
    jnc @F
    mov al,'.'
@@:
    push ecx
    invoke printf, CStr("%c"), eax
    pop ecx
    loop nextc
    invoke printf, CStr(lf)
    pop ecx
    loop nextline
    mov [address],esi
    ret
d_cmd endp

;--- 'q': back to real-mode

q_cmd proc
    mov ax,4c00h
    int 21h
q_cmd endp

;--- set exception 0E handler so we 
;--- won't terminate unexpectedly

set_exception_handlers proc

    mov ecx, cs
    mov edx, offset exception0E
    mov bl, 0Eh
    mov ax, 203h
    int 31h
    ret

set_exception_handlers endp

;--- Dos32cm provides a 64-bit exception frame!

exception0E:
    sti
    mov edx, [esp+0*8]
    mov ecx, [esp+1*8]
    mov eax, [esp+2*8]
    mov ebx, cr2
    invoke printf, CStr(lf,"page fault, errcode=%X cs:eip=%X:%X cr2=%X",lf), edx, eax, ecx, ebx
    mov esp, dwEsp
    ret

    end main


Mon32X/MON32x.asm

;--- mixing 32- and 64-bit sections in ".model flat"
;--- this requires JWasm v2.17 and using option -pe.

    .386
    .model flat
    option casemap:none
    option proc:private

TCMD equ 0      ; 1=activate 't'

lf  equ 10 

    .data?

qwRsp   dq ?    ; saved stack pointer
address dq ?    ; current offset for d cmd

buffer db 20 dup (?); keyboard buffer
lbuffer equ $ - offset buffer

    .code

f64 label fword
    dd offset main64
    dw 8        ; selector 8 is 64-bit code segment ( Dos32cm specific )

main proc c

    jmp [f64]

main endp

;--- define a string
CStr macro text:vararg
local sym
    .const
sym db text,0
    .code _TEXT64
    exitm <addr sym>
endm

    .x64

_TEXT64 segment use64 'CODE'
_TEXT64 ends

    .code _TEXT64

    include printf.inc

main64 proc

    and sp, 0fff0h
    invoke printf, CStr("Mon32x loaded at %X, rsp=%lX",lf), rbx, rsp
    invoke printf, CStr("ax=%X bx=%X cx=%X dx=%X si=%X di=%X bp=%X",lf), rax, rbx, rcx, rdx, rsi, rdi, rbp
    call set_exception_handlers
nextcmd:
    invoke printf, CStr("(cmds: a,d,q): ")

    mov ah,1        ; read a key from keyboard with echo
    int 21h
    lea rcx, nextcmd
    push rcx
    mov [qwRsp], rsp
    push rax
    invoke printf, CStr(lf)
    pop rax
    cmp al,'a'
    jz a_cmd
    cmp al,'d'
    jz d_cmd
    cmp al,'q'
    jz q_cmd
if TCMD
    cmp al,'t'
    jz t_cmd
endif
    cmp al,0dh      ;ENTER?
    jz newline
    mov ecx, eax
    invoke printf, CStr("unknown cmd: %c",lf), rcx
newline:
    ret
main64 endp

;--- get a line of characters
;--- ebx->buffer
;--- esi=size of buffer

getline proc
    xor edi, edi
    dec esi
nextkey:
    mov ah,1
    int 21h
    cmp al,0dh
    jz enter_pressed
    cmp al,08h
    jz backspace_pressed
    mov [ebx+edi],al
    inc edi
    cmp edi,esi
    jnz nextkey
    mov dl,0dh
    mov ah,2
    int 21h
    jmp enter_pressed
backspace_pressed:
    mov dl,20h
    mov ah,2
    int 21h
    cmp edi,0
    jz nextkey
    dec edi
    mov dl,08h
    mov ah,2
    int 21h
    jmp nextkey
enter_pressed:    
    mov byte ptr [ebx+edi],0
    ret
getline endp

;--- enter address (for d cmd)

a_cmd proc
    invoke printf, CStr("enter start address for d cmd: ")
    mov esi, lbuffer
    lea rbx, buffer
    call getline
    and edi,edi        ;at least 1 digit entered?
    jz done
    xor edi,edi
    xor rsi,rsi
    .while byte ptr [rbx+rdi]
        mov al,byte ptr [rbx+rdi]
        sub al,'0'
        jc error
        cmp al,9
        jbe @F
        or al, 20h
        sub al, 27h
        cmp al, 0fh
        ja error
@@:
        movzx eax,al
        shl rsi,4
        add rsi,rax
        inc rdi
    .endw
;--- check if address valid
    mov rax, rsi
    shr rax, 48-1
    jz @F              ; ok if bits 47-63 are all 0
    cmp rax, 1ffffh
    jnz error          ; ok if bits 47-63 are all 1
@@:
    mov [address],rsi
done:
    invoke printf, CStr(lf)
    ret
error:
    invoke printf, CStr(lf,"%s?",lf), addr buffer
    ret

a_cmd endp

;--- display memory dump

d_cmd proc
    mov rsi, [address]
    mov ecx, 8
nextline:
    push rcx
    invoke printf, CStr("%08lX: "), rsi
    mov ecx, 16
@@:
    push rcx
    lodsb
    movzx eax, al
    invoke printf, CStr("%02X "), eax
    pop rcx
    loop @B
    invoke printf, CStr(" ")
    sub rsi, 16
    mov ecx, 16
nextc:
    lodsb
    cmp al, 20h
    jnc @F
    mov al, '.'
@@:
    push rcx
    invoke printf, CStr("%c"), rax
    pop rcx
    loop nextc
    invoke printf, CStr(lf)
    pop rcx
    dec ecx
    jnz nextline
    mov [address], rsi
    ret
d_cmd endp

;--- 'q': back to real-mode

q_cmd proc
    mov ax,4c00h
    int 21h
q_cmd endp

if TCMD

;--- 't': some printf tests

t_cmd proc
    or rax, -1
    invoke printf, CStr("%u %d %x",10), rax, rax, rax
    or rax, -1
    invoke printf, CStr("%lu %ld %lx",10), rax, rax, rax
    ret
t_cmd endp

endif

;--- handle exception 0E so we won't terminate
;--- if an invalid address has been entered.

set_exception_handlers proc

    mov ecx, cs
    mov edx, offset exception0E
    mov bl, 0Eh
    mov ax, 203h
    int 31h
    ret

set_exception_handlers endp

;--- Dos32cm provides a 64-bit exception frame!

exception0E:
    sti
    mov rdx, [rsp+0*8]
    mov rcx, [rsp+1*8]
    mov rax, [rsp+2*8]
    mov rbx, cr2
    invoke printf, CStr(lf,"page fault, errcode=%X cs:eip=%X:%X cr2=%lX",lf), rdx, rax, rcx, rbx
    mov rsp, [qwRsp]
    ret

    end main
